easybuild/tools/filetools.py

```diff
@@ -569,6 +569,70 @@ def derive_alt_pypi_url(url):
+    """
+    Recurse into [URLPAT::][HEADER: ]FILE|FIELD where FILE may be another such string or
+    file containing lines matching the same format, and flatten the result as a dict
+    e.g. {'^example.com': ['Authorization: Basic token', 'User-Agent: Special Agent']}
+    """
+    # stop infinite recursion that might happen if a file.txt refers to itself
+    if maxdepth < 0:
+        _log.debug("Failed to parse_http_header_fields_urlpat (recursion limit)")
+        return urlpat_headers
+
+    if not isinstance(arg, str):
+        _log.debug("Failed to parse_http_header_fields_urlpat (argument not a string)")
+        return urlpat_headers
+
+    # HTTP header fields are separated by CRLF but splitting on LF is more convenient
+    for argline in arg.split('\n'):
+        argline = argline.strip()  # remove optional whitespace (e.g. remaining CR)
+        if argline == '' or '#' in argline[0]:
+            continue  # permit comment lines: ignore them
+
+        if os.path.isfile(os.path.join(os.getcwd(), argline)):
+            # expand existing relative path to absolute
+            argline = os.path.join(os.path.join(os.getcwd(), argline))
+        if os.path.isfile(argline):
+            # argline is a file path, so read that instead
+            _log.debug('File included in parse_http_header_fields_urlpat: %s' % argline)
+            argline = read_file(argline)
+            urlpat_headers = parse_http_header_fields_urlpat(argline, urlpat, header, urlpat_headers, maxdepth - 1)
+            continue
+
+        # URL pattern is separated by '::' from a HTTP header field
+        if '::' in argline:
+            [urlpat, argline] = argline.split('::', 1)  # get the urlpat
+            # the remainder may be another parseable argument, recurse with same depth
+            urlpat_headers = parse_http_header_fields_urlpat(argline, urlpat, header, urlpat_headers, maxdepth)
+            continue
+
+        # Header field has format HEADER: FIELD, and FIELD may be another parseable argument
+        # except if FIELD contains colons, then argline is the final HEADER: FIELD to be returned
+        if ':' in argline and argline.count(':') == 1:
+            [argheader, argline] = argline.split(':', 1)  # get the header and the remainder
+            # the remainder may be another parseable argument, recurse with same depth
+            # note that argheader would be forgotten in favor of the urlpat_headers returned by recursion,
+            # so pass on the header for reconstruction just in case there was nothing to recurse in
+            urlpat_headers = parse_http_header_fields_urlpat(argline, urlpat, argheader, urlpat_headers, maxdepth)
+            continue
+
+        if header is not None:
+            # parent caller didn't want to forget about the header, reconstruct as recursion stops here.
+            argline = header.strip() + ': ' + argline
```

Comment:
this is introducing a space in the value that is not stripped in `download_file`

in order for the tests to catch these problems it may be better for the test to directly call and assert the output of  `parse_http_header_fields_urlpat`

Reply:
I wanted to add that space initially, since it would be stripped off from `argline`, but since [that white-space is optional](https://tools.ietf.org/html/rfc7230#section-3.2), I'll remove it.

I'll try to add some direct tests on function `parse_http_header_fields_urlpat`. In fact, you may prefer testing this over the testing of `download_file` (many times) as the latter will needlessly try downloading a gzip that many times.